Explore/Logs: Improve and add descriptions of Explore's Escape newlines feature

gguillotte-grafana opened the pull request:
**What this PR does / why we need it**:

Per issue #45770, the tooltip describing the "Escape newlines" feature in Explore suggests the feature is experimental, which can be interpreted to mean the feature might be removed. The intent was to communicate to users that the results should be manually confirmed, since the automatic replacement of incorrectly escaped sequences might not always be accurate.

This PR rewrites the tooltip to clarify these points, and adds documentation describing the feature.

**Which issue(s) this PR fixes**:

Fixes #45770 

**Special notes for your reviewer**:

This PR changes a tooltip in code and also adds supporting documentation. Please let me know if these changes should be split into separate PRs for the code change and docs addition.

gguillotte-grafana committed:
Explore/Logs: Clarify phrasing of newline escape fix tooltip

Rewrite the tooltip for the smart feature in PR #31352 that
replaces incorrectly escaped newlines in log lines.

-   Clarify the functionality of the feature to emphasize its
    interactivity.
-   Remove language suggesting that the feature is experimental,
    when we intended to suggest reviewing the results manually for
    correctness.

gguillotte-grafana committed:
Docs: Document escape newlines feature

Describe and provide steps for using the "Escape newlines" feature
in Explore added in PR #31352.

ivanahuckova commented:
This is awesome! Thanks @gguillotte-grafana for tackling this.

achatterjee-grafana commented on public/app/features/explore/LogsMetaRow.tsx:
```diff
@@ -71,7 +71,7 @@ export const LogsMetaRow: React.FC<Props> = React.memo(
         value: (
           <Tooltip
-            content="We suggest to try to fix the escaping of your log lines first. This is an experimental feature, your logs might not be correctly escaped."
+            content="Attempt to automatically fix incorrectly escaped newline and tab sequences in log lines. Manually review the results to confirm that the replacements are correct."
```
```suggestion
            content="Fix incorrectly escaped newline and tab sequences in log lines. Manually review the results to confirm that the replacements are correct."
```

gguillotte-grafana replied:
Thanks! I waffled on whether to be this explicit about the feature's reliability.

achatterjee-grafana replied:
"Attempt to automatically fix" does not read well, it can also reduce user confidence in the feature capability.

The preceding label says: "'Your logs might have incorrectly escaped content'. So, perhaps say: "Automatically fix incorrectly ..."

GrafanaWriter replied:
I like @achatterjee-grafana's edit here.

achatterjee-grafana commented on docs/sources/explore/logs-integration.md:
```diff
@@ -97,6 +97,16 @@ You can change the order of received logs from the default descending order (new
+
+> **Note:** Available in Grafana 7.5 and later versions.
+
+Explore can automatically detect and attempt to fix incorrectly escaped sequences in log lines, such as newlines (`\n`, `\r`) or tabs (`\t`). This feature appears only when Explore detects incorrectly escaped sequences.
```
```suggestion
Explore can automatically detect and attempt to fix incorrectly escaped sequences in log lines, such as newlines (`\n`, `\r`) or tabs (`\t`). 
```

gguillotte-grafana replied:
This feature won't be visible in many circumstances. I'm concerned that omitting the conditional nature (and concerns about reliability that led to calling it "experimental") could confuse users who expect it to be visible. Is there a better way to emphasize this?

achatterjee-grafana replied:
How about "The "Escape newlines" option is available only when Explore detects incorrectly escaped sequences."

achatterjee-grafana replied:
Also, this capability is probably no longer "experimental". Might be worth it to find out more.

gguillotte-grafana replied:
> Also, this capability is probably no longer "experimental". Might be worth it to find out more.

The linked issue has some discussion about this. "Experimental" here referred to the relative reliability of detection and replacement, not the supportability of the feature.

> I would actually remove the note "Available in Grafana 7.5 and later versions".

👍🏻

achatterjee-grafana commented on docs/sources/explore/logs-integration.md:
```diff
@@ -97,6 +97,16 @@ You can change the order of received logs from the default descending order (new
+
+Explore can automatically detect and attempt to fix incorrectly escaped sequences in log lines, such as newlines (`\n`, `\r`) or tabs (`\t`). This feature appears only when Explore detects incorrectly escaped sequences.
+
+1. Click the "Escape newlines" button to fix incorrectly escaped sequences.
```
```suggestion
1. Click the "Escape newlines" option to fix incorrectly escaped sequences.
```

achatterjee-grafana commented on docs/sources/explore/logs-integration.md:
```diff
@@ -97,6 +97,16 @@ You can change the order of received logs from the default descending order (new
+Explore can automatically detect and attempt to fix incorrectly escaped sequences in log lines, such as newlines (`\n`, `\r`) or tabs (`\t`). This feature appears only when Explore detects incorrectly escaped sequences.
+
+1. Click the "Escape newlines" button to fix incorrectly escaped sequences.
+2. Manually review the replacements to confirm that the detection and replacement were correct.
```
```suggestion
2. Manually review the replacements to confirm that the detection and replacement are correct.
```

achatterjee-grafana commented on docs/sources/explore/logs-integration.md:
```diff
@@ -97,6 +97,16 @@ You can change the order of received logs from the default descending order (new
+
+1. Click the "Escape newlines" button to fix incorrectly escaped sequences.
+2. Manually review the replacements to confirm that the detection and replacement were correct.
+3. To revert the replacements, click "Remove escaping".
```
```suggestion
3. If you need to revert the replacements, then click "Remove escaping".
```

gguillotte-grafana replied:
Reverting should be a separate (if one-step) task, and this should also mention that the option changes, so I'll rewrite.

achatterjee-grafana reviewed (commented):
Added doc review.

gguillotte-grafana committed:
Rewrite topic lead to clarify conditional behavior

gguillotte-grafana committed:
Describe reversion of replacements as a standalone task

Reverting the replacements is a separate action, so it should have
its own task. This also describes the option's transformation and
tightens the task language.

gguillotte-grafana committed:
Remove Grafana 7 version qualifier

Elfo404 commented:
@achatterjee-grafana can you take a look please?

achatterjee-grafana commented on docs/sources/explore/logs-integration.md:
```diff
@@ -97,6 +97,17 @@ You can change the order of received logs from the default descending order (new
+1. Click "Escape newlines" to replace the sequences.
+2. Manually review the replacements to confirm their correctness.
+
+After replacing these sequences, Explore changes the option from "Escape newlines" to "Remove escaping". Clicking this option reverts the replacements.
```
```suggestion
If you want to revert the replacements, then click the "Remove escaping" option.
```

gguillotte-grafana replied:
Because this involves a UI element changing, and not two separate UI elements, the "Remove escaping" option won't always be visible. This suggestion would mean the conditions for that UI element's presence won't be documented. Is this intended?

GrafanaWriter replied:
I've added an edit to see if we can make that more clear.

achatterjee-grafana reviewed (commented):
Added doc review.

Elfo404 commented:
@grafana/observability-logs-and-traces do you mind having a second look at this? AFAIK we do not guarantee the correctness of the replacements. everything else looks good on our side, i just want to confirm we are communicating the correct information to users

gguillotte-grafana commented:
@achatterjee-grafana @grafana/observability-logs-and-traces This PR was meant to solve #45770, which specifically raised concerns about managing user expectations for this feature's reliability, and that the goal is only to remove the confusing usage of "experimental", not to document that the feature is reliable.

My understanding is that we won't guarantee that this feature will correctly fix all escaping issues, and I'm concerned that the suggested changes won't communicate this.

Emphasis mine:

> Of course, it might always produce incorrect results, as we don't know which characters are actual newlines and which have just the \n pattern, but I think that *if we communicate it clearly, users will understand.*

If we don't want to communicate this, perhaps we should close this, discuss in #45770 what it is that we want here instead, and I can rewrite and open a new PR.

achatterjee-grafana commented:
It is tricky documenting something that may or may not work - I agree with @gguillotte-grafana, we should revisit how we mention this in our docs.

Also, ping @GrafanaWriter, for your opinion.

GrafanaWriter commented on docs/sources/explore/logs-integration.md:
```diff
@@ -97,6 +97,17 @@ You can change the order of received logs from the default descending order (new
+1. Click "Escape newlines" to replace the sequences.
+2. Manually review the replacements to confirm their correctness.
+
+After replacing these sequences, Explore changes the option from "Escape newlines" to "Remove escaping". Clicking this option reverts the replacements.
```
```suggestion
 Explore replaces these sequences. When it does so, the option will change from "Escape newlines" to "Remove escaping". Evaluate the changes as the parsing may not be accurate based on the input received. You can revert the replacements by clicking "Remove escaping".
```

gguillotte-grafana replied:
Removed an extraneous leading space before committing. Thank you!

GrafanaWriter reviewed (approved):
Thank you.

gguillotte-grafana committed:
Clarify escape sequence detection lede on task

Co-authored-by: Fiona Artiaga <89225282+GrafanaWriter@users.noreply.github.com>

gguillotte-grafana committed:
Clarify "Remove escaping" state change

Co-authored-by: Fiona Artiaga <89225282+GrafanaWriter@users.noreply.github.com>

gguillotte-grafana committed:
Clarify confidence of tooltip content

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

gguillotte-grafana commented:
Thanks @achatterjee-grafana and @grafanawriter! I've committed the changes.

gguillotte-grafana commented:
Anything else I need to do to move this forward?

gguillotte-grafana commented:
I don't have merge permissions so I do still need someone to merge this PR.

ivanahuckova commented:
merged 🙂
